src/v/cloud_storage/inventory/ntp_hashes.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "model/fundamental.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/file.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/iostream.hh>
+
+#include <absl/container/node_hash_set.h>
+
+namespace cloud_storage::inventory {
+
+enum class lookup_result {
+    exists,
+    missing,
+    possible_collision,
+};
+
+class ntp_path_hashes {
+public:
+    ntp_path_hashes(model::ntp ntp, std::filesystem::path hash_store_path);
+
+    ss::future<bool> load_hashes();
+
+    lookup_result exists(const remote_segment_path& path) const;
+
+    ss::future<> stop();
+
+private:
+    ss::future<> load_hashes(ss::directory_entry de);
+    ss::future<> load_hashes(ss::input_stream<char>& stream);
+
+    model::ntp _ntp;
+    std::filesystem::path _ntp_hashes_path;
+    absl::node_hash_set<uint64_t> _path_hashes;
+    absl::node_hash_set<uint64_t> _possible_collisions;
```

Comment:
Just want to make sure I'm understanding tradeoff we're making: by allowing collisions with a smaller hash representation, we're giving up on doing a 100% accurate job scrubbing (compared if we used more hash bits or a cstore to store the segments)

Are we thinking the tradeoff is fine because the filtering by NTP also gives us an added layer of collision avoidance, where it's even less likely multiple hashes within a single NTP would collide?

Reply:
The scrubber is expected to make an actual HEAD request if a collision is detected for a path, so we might end up making extra HTTP calls, one per collision, but we wouldn't have to give up on accuracy. The scrubber has the following directives:

* it finds a path in the collisions map - make a HEAD request to check
* it finds a path in hashes map but not in collisions - path exists
* it finds path in neither map - make a HEAD request to check

The separate hash set for tracking collisions gives us the ability to not compromise on accuracy by treating collisions as potentially missing segments.

Reply:
>but we wouldn't have to give up on accuracy

Sorry just following up on this now, but I found myself thinking about this again in the reviewing a more recent PR. Let's take an example:

- string "segment-1" hashes to hash-a
- string "segment-2" also hashes to hash-a
- "segment-1" exists and "segment-2" doesn't exist in the bucket, so the inventory hashes include hash-a
- our manifest contains "segment-2" so when we look for "segment-2" it looks up hash-a
- we return that hash-a exists, even though segment-2 doesn't exist

Is this scenario avoided somehow? Is it by having a good enough hash that we don't actually expect collisions?

I'm thinking that we expect that if this were the case, segment-1 and segment-2 would both exist in a manifest somewhere, and therefore we'd be able to detect the collision. But it seems theoretically possible for a poor hash function to result in a scenario where "segment-1" is some orphaned segment and we incorrectly think that segment-2 exists when it doesn't.

Reply:
This scenario should be handled, it is not by the properties of the hash itself, but because of the data structure (vector) we use to store the hashes during write. We can walk through the example scenario, let me know if something seems incorrect or missing. 

As a background, we manage path hashes in two distinct phases:

* the "write" phase, which is inventory consumer,  run through inventory service, culminating in files on disk containing the 8 byte hashes
* followed by the "read" phase, where the hash loader reads through the files and creates in-memory sets.

If we consider the scenario you mentioned, focusing on the write phase first:

* string "segment-1" hashes to hash-a. hash-a is stored in the vector for the NTP, in memory, by the inv-consumer
* string "segment-2" also hashes to hash-a. hash-a is also stored in the same vector. this step is crucial and enables us to handle collisions - we are storing hashes in a frag. vector in memory. we do not create sets yet, so there are two instances of hash-a at different indices in the vector, we are not replacing anything, just adding each new hash to the vector
* eventually data is flushed to disk, it is a serialized frag. vector, this means that there are still two instances of hash-a on disk. they could be in the same file or in different files, but always within the NTP directory where the hash loader looks for them.
* when the write phase finishes the hash file(s) contains two instances of hash-a because of collision. this is the case for every collision, there are `>1` instances of colliding hash number in the files which contain vectors.

So we have preserved a clue about the collision in the write phase, which is that a hash occurs twice (or more than twice) in the final data.

When the read phase begins:
* the loader loads up the frag. vectors from disk and starts merging them into a main set, but before adding each hash to the main set, it checks if the hash was already there
* if the hash was already there, as will happen with hash-a, because it appears twice, then we know at this point there was a collision
* we add hash-a to the special collision set when we see it was already in the main set
* the set population finishes, with all "good" hashes in the main set, and all detected collisions in the collision set. note that hash-a can be in both sets, or just collision set. either case is fine.

Now the scrubber queries the loader. If the hash for a path is found in the collision set (which we check first), then this is a signal to the scrubber not to trust the dataset and perform a conservative check which is a HEAD request. 

* when segment-a is hashed during the query, we get hash-a. this is found in the collision set, so the scrubber will do a HEAD request for segment-a, not trusting the data from inventory hashes
* when segment-b is hashed, again the result is hash-a, so another HEAD request will be made.
* As a result of the HEAD requests we know if there is a missing segment.

This completes the process, we have to make extra HEAD requests for each possible collision. 

There is an obvious trade-off here. The larger the number of possible collisions, the bigger the collision set in the hash loader, and more HEAD requests have to be made. As long as the collisions are below a manageable threshold this should result in very few HEAD calls required. 

Note that we can swap out the hash algorithm and still keep the same data structures to manage collisions, for example if we switch to 128 bit hash we will get much fewer collisions at the expense of larger memory and disk usage. We could even swap out to xxhash32 for smaller data sets.

But following the above algorithm, we always have the guarantee of not getting a false positive as long as we track the collisions and do a conservative check when a collision is detected.

Let me know if the above makes sense or there is something I've overlooked.

Reply:
>string "segment-2" also hashes to hash-a. hash-a is also stored in the same vector. this step is crucial and enables us to handle collisions

Thanks for the detailed response, but I think I'm still missing something. In my example, segment-2 doesn't exist in the bucket, and so it wouldn't be consumed by the inv-consumer.

I mention segment-2 to emphasize that hashes aren't necessarily unique. If you can construct two valid segments names that hash to the same value, you wouldn't be able to tell with the hash alone whether segment-1 exists or segment-2. If segment-1 exists but segment-2 doesn't, a test for the existence of segment-2 would say that it exists if we only use hashes, no?

Reply:
I see, the main issue is this (which I missed in my original reading):

> "segment-1" exists and "segment-2" doesn't exist in the bucket, so the inventory hashes include hash-a

So we have hash-a, but it comes from a random path in the report we don't care about, and the segment we do care about is also hash-a, but is not in the bucket. Yeah, that would be a false positive. 

As it stands we are only able to detect hash collisions when multiple segments produced the same hashes and then one of them was deleted, our correctness depends on there being multiple hashes to signal a collision. But the case you described would be missed.

We can switch the hash algorithm  (with not much effort) to a 128 bit hash, to reduce the possibility, but not completely remove it.  But if we have just one hash, there is always a risk that it came not from the segment we expect but from some other item in the inventory report like you described.

Reply:
One relatively cheap way to reduce this false positive even further than hash collisions is to keep a bitmap of the start offsets of segments seen per NTP while consuming the report. 

Because start offsets are somewhat unique (barring segment reupload/overlap anomaly) per NTP, if when querying the hash set, we also then query the bitmap for the start offset of the segment in manifest, then this decreases the chance of running into a false positive. 

In this case segment-2 which was deleted, but is in manifest and has hash-a, will have a missing start offset in the bitmap for the NTP (unless it was particularly unlucky and another segment also had the same offset due to another anomaly or reupload). 

This still doesn't give a 100% accuracy though.